src/main/scala/mesosphere/marathon/state/Volume.scala

```diff
@@ -17,106 +23,251 @@ object Volume {
     containerPath: String,
     hostPath: Option[String],
     mode: Mesos.Volume.Mode,
-    persistent: Option[PersistentVolumeInfo]): Volume =
-    persistent match {
-      case Some(persistentVolumeInfo) =>
-        PersistentVolume(
-          containerPath = containerPath,
-          persistent = persistentVolumeInfo,
-          mode = mode
-        )
-      case None =>
-        DockerVolume(
-          containerPath = containerPath,
-          hostPath = hostPath.getOrElse(""),
-          mode = mode
-        )
-    }
+    persistent: Option[PersistentVolumeInfo],
+    external: Option[ExternalVolumeInfo]): Volume = {
 
-  def apply(proto: Protos.Volume): Volume = {
-    val persistent: Option[PersistentVolumeInfo] =
-      if (proto.hasPersistent) Some(PersistentVolumeInfo(proto.getPersistent.getSize)) else None
-
-    persistent match {
-      case Some(persistentVolumeInfo) =>
-        PersistentVolume(
-          containerPath = proto.getContainerPath,
-          persistent = persistentVolumeInfo,
-          mode = proto.getMode
-        )
-      case None =>
-        DockerVolume(
-          containerPath = proto.getContainerPath,
-          hostPath = proto.getHostPath,
-          mode = proto.getMode
-        )
+    if (persistent.isDefined) {
```

Comment:
this feels weak! what stops someone from defining both persistent and external in the same volume spec?